Re-attach SRIOV devices to target VM when live migration fails or is canceled

ormergi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
When migrating a VM with SRIOV host-devices is aborted due to failure or cancel request,
the VM will end up with no SRIOV devices connected until the next reboot.

The reason for that is the fact that we detach SRIOV host-devices just before migration starts 
due to Libvirt won't allow migrating VM’s with connected VFIO devices (e.g: SRIOV VF).

To overcome this, SRIOV host-devices will be re-attached to the VM on source when a migration fails or is canceled.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Now SRIOV host-devices re-connects to VM on source when migration is aborted
```

ormergi commented:
/cc @EdDev

ormergi commented:
rebased

oshoval commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -332,6 +337,11 @@ func (l *LibvirtDomainManager) cancelMigration(vmi *v1.VirtualMachineInstance) e
+
+	log.Log.Object(vmi).Info("migration is aborted, reconnecting sriov host-devices")
+	if reconnectSriovDevicesError := l.hotPlugHostDevices(vmi); err != nil {
+		log.Log.Object(vmi).Reason(reconnectSriovDevicesError).Error("failed to hot-plug host-devices")
```
shouldn't it return err here if the reconnect failed?

EdDev reviewed (commented):
> tests, sriov: Refactor migration test

This commit is missing the "why" part. I do not understand why the migration is now triggered in the fixture and not in the test body.
My understanding was that the test is checking the migration itself, not only its end result.

EdDev commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -267,6 +267,11 @@ func (l *LibvirtDomainManager) startMigration(vmi *v1.VirtualMachineInstance, op
 		log.Log.Object(vmi).Reason(err).Error("Live migration failed.")
 		l.setMigrationResult(vmi, true, fmt.Sprintf("%v", err), "")
+
+		log.Log.Object(vmi).Info("migration failed, reconnect sriov host-devices")
+		if reconnectSriovDevicesError := l.hotPlugHostDevices(vmi); err != nil {
+			log.Log.Object(vmi).Reason(reconnectSriovDevicesError).Error("failed to hot-plug host-devices")
+		}
```
This whole error block seems redundant/no-op to me.
`asyncMigrate` will always return `nil`, there is nothing there to return an error.

ormergi replied:
Thanks for the head-up, you right I will remove it

EdDev reviewed (commented):
commit:  tests, migration: Refactor RunStressTest

> The second argument than being passed to 'stress' tool --vm-bytes argument is expected to be a number w/o unit suffix.

"than" seems not to fit here.

> Add existence check for 'stress' binary before running it in order to prevent unexpected behavior.

I see you used `which` embedded in the shell command. I do not understand why it is needed.
If the binary is missing, it will give a relevant error, so why do this addition?

ormergi commented:
> commit: tests, migration: Refactor RunStressTest
> 
> > The second argument than being passed to 'stress' tool --vm-bytes argument is expected to be a number w/o unit suffix.
> 
> "than" seems not to fit here.
 
Thanks for the heads up I will remove it

> > Add existence check for 'stress' binary before running it in order to prevent unexpected behavior.
> 
> I see you used `which` embedded in the shell command. I do not understand why it is needed.
> If the binary is missing, it will give a relevant error, so why do this addition?

Due to stress tool is invoked as background job if the binary it missing the error will not propagate

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -2356,6 +2357,7 @@ func (d *VirtualMachineController) vmUpdateHelperMigrationSource(origVMI *v1.Vir
 
 		err = client.MigrateVirtualMachine(vmi, options)
 		if err != nil {
+			d.finalizeMigration(vmi)
```
If the code reaches this block, I do not understand why there is a need to call finalize. It seems to me like this may occur only as a setup failure, to the migration. But if that occurs, finalize should not be called.
Finalize should be called only if the migration already started.

If you remove this line, are there e2e tests which fail? (in other words, do you manage to cover this case?)

ormergi replied:
> If the code reaches this block, I do not understand why there is a need to call finalize. 
It seems to me like this may occur only as a setup failure, to the migration. But if that occurs, finalize should not be called.
> Finalize should be called only if the migration already started.

If the migration fails due to setup failure or while running, we will get to this line.

The way I see it `finalizeMigration` should be called whether a migration if failed or not,
its the logic to run when the migration is finished no matter if it failed or not.
Currently there is no clear point in the code that handle the part when the migration is finished - post migration.
As you may already know, there are multiple points on virt-handler code that indicates the migration is finished,
where on each point the migration state is different state.
With the current implementation I am not sure if its possible to have such place in the code
w/o changing `virt-controller` and `virt-handler` migration logic, which is in my opinion too much work for now.
On the other hand if we had such point, we could call `finalizeMigration` once and not needing this PR at all.
> 
> If you remove this line, are there e2e tests which fail? (in other words, do you manage to cover this case?)

Yes, both tests of the failing path fails:
[Failure while migration is running test](https://github.com/kubevirt/kubevirt/pull/5922/commits/f73036acf559f4cf05f9e1de15e44d1a0b6be667#diffda14ca97e77db6bb275427ac69187a539c58fc297e64519628e5385f9f411c5fR1049)
[Failure during setup test ](https://github.com/kubevirt/kubevirt/pull/5922/commits/a5e55d3aebbaa81a4d80f028a4466b012c4652f1#diff-da14ca97e77db6bb275427ac69187a539c58fc297e64519628e5385f9f411c5fR1024)

EdDev replied:
I'm following the code here, let me know when I missed something:
- `client.MigrateVirtualMachine` call reaches the [virt-launcher](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/cmd-server/server.go#L96).
- It calls [MigrateVmi](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/manager.go#L370) which calls in turn to [startMigration](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L248).
- All these do some preparations, but do not really detach or prepare the domain for migration. The interesting stuff happen in [asyncMigrate](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L883) but this one will never return an error, so failure in detachment for example will not be propagated back. Nor the artificial assertion with the [annotation](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L887).

ormergi replied:
You right asyncMigrate will not return anything (the fact that on the caller there is error handling is bad)
But the errors for the detachment and the artificial assertion with the annotation will propagate to virt-handler.

If an error accrues during asyncMigrate, for example:
detachment fails 
- The caller of detachment is [migrationHelper](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L806),
returns an error that is passed to the error-channel the migration monitor listens to [[1]](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L893-L910) [[2]](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L623) [[3]](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L657-L660) 

- migration monitor will updated the domain migration metadata [[4]](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L668).

- vm-controller on virt-handler, updates the VMI `Status.MigrationState` according to the domain migration meta [setMigrationProgressStatus](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-handler/vm.go#L573)
- Now the migraiton failure is reflected on the VMI status that will reconciled by the virt-handler

EdDev replied:
You explained the flow correctly per what I know, but I think the connection to calling `client.MigrateVirtualMachine(vmi, options)` is lost here.
The call to that function will not return an error due to a detachment, migration or other setup failures, it will just cause what you just mentioned and get reflected at some point in the future in the VMI status.
If `client.MigrateVirtualMachine(vmi, options)` returns an error, it means we never reached the setup or the migration at all, we failed much earlier. So why would we call finalize then?

And with that the obvious question raises: If a failure occurs in the setup (at the detach, at the annotation condition, etc) or in the middle of the migration, and the error is populated to the VMI status eventually, why is the filizeMigration called and when.

ormergi replied:
I checked it more thoroughly
Its like you said, when `client.MigrateVirtualMachine` is called we wont catch errors from 
`asyncMigrate`(gorutine) and forwarder including: detachment, migration or other setup failures.

We do catch errors form until `asyncMigrate` is called, for example:
when a migration is already running will get "migration job already executed" error.
The error will propagate and `finalizeMigration` will run.
Here is the path:
`client.MigrateVirtualMachine` ->  `MigrateVirtualMachine` -> `migrateVMI` [[1]](https://github.com/kubevirt/kubevirt/blob/a5e55d3aebbaa81a4d80f028a4466b012c4652f1/pkg/virt-launcher/virtwrap/cmd-server/server.go#L110)  -> 
`startMigration` [[2]](https://github.com/kubevirt/kubevirt/blob/a5e55d3aebbaa81a4d80f028a4466b012c4652f1/pkg/virt-launcher/virtwrap/manager.go#L370) [[3]](https://github.com/kubevirt/kubevirt/blob/a5e55d3aebbaa81a4d80f028a4466b012c4652f1/pkg/virt-launcher/virtwrap/live-migration-source.go#L248) -> `initializeMigrationMetadata` [[4]](https://github.com/kubevirt/kubevirt/blob/a5e55d3aebbaa81a4d80f028a4466b012c4652f1/pkg/virt-launcher/virtwrap/live-migration-source.go#L254) 
and finally we will hit [[5]](https://github.com/kubevirt/kubevirt/blob/a5e55d3aebbaa81a4d80f028a4466b012c4652f1/pkg/virt-launcher/virtwrap/live-migration-source.go#L302) 

Also this code run more then once (as it invoked from vm controller loop),
which means we will always get these errors.
This explains why [failure while migration is running test](https://github.com/kubevirt/kubevirt/blob/f73036acf559f4cf05f9e1de15e44d1a0b6be667/tests/network/vmi_multus.go#L1049) pass's.

EdDev reviewed (commented):
Partial review of the "sriov migration: Reattach sriov devices when aborted" commit.

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -2341,6 +2341,7 @@ func (d *VirtualMachineController) vmUpdateHelperMigrationSource(origVMI *v1.Vir
 				return err
 			}
 			d.recorder.Event(vmi, k8sv1.EventTypeNormal, v1.Migrating.String(), "VirtualMachineInstance is aborting migration.")
+			d.finalizeMigration(vmi)
```
If I read this right, the `finalizeMigration` is to be executed when `AbortRequested`, therefore this is done when the virt-controller updates the flag.

First, I am not sure if we may pass though this code flow more than once.
Based on the [cancel function](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L322) at the virt-launcher, this portion of the code is only relevant to explicitly abort a running migration and not to do so on an already failed one. Your addition now changes this.

We have these scenario I know of:
- Setup of the migration failed (migration was not started yet on the domain itself).
- Migration failed, sourcing from libvirt itself.
- Migration failed, sourcing from the monitoring go-routine at the source node.
- Migration failed because the virt-controller decided so (or because a user asked to cancel).

All of these are handled here?
For example, what happens if the monitor marks the migration as failed, without an abort reason? Is that propagated to the virt-handler and then to the virt-controller which in turn sets this `AbortRequested` to `true`?

ormergi replied:
@EdDev 
> If I read this right, the `finalizeMigration` is to be executed when `AbortRequested`, therefore this is done when the virt-controller updates the flag.
> 

True

> First, I am not sure if we may pass though this code flow more than once.

If you mean that due to an VMI update issue we will get there again then, yes this is what going to happen.

Upper the call stack, `defaultExecute` call's this function and eventually updates the VMI status as part of VMI sync.
In case of an error the request will re-queued and this code will run again.

> Based on the [cancel function](https://github.com/kubevirt/kubevirt/blob/099be903803556bfe3a85075a7b55f0a711d9ca7/pkg/virt-launcher/virtwrap/live-migration-source.go#L322) at the virt-launcher, this portion of the code is only relevant to explicitly abort a running migration and not to do so on an already failed one. Your addition now changes this.
> 
What does it changes? you think its not contradicts with cancel logic on virt-launcher?

> We have these scenario I know of:
> 
> * (1) Setup of the migration failed (migration was not started yet on the domain itself).
> * (2) Migration failed, sourcing from libvirt itself.
> * (3) Migration failed, sourcing from the monitoring go-routine at the source node.
> * (4) Migration failed because the virt-controller decided so (or because a user asked to cancel).
> 
> All of these are handled here?

(4) - handled here.
(1), (2) and (3) - are handled at line  [[1]](https://github.com/kubevirt/kubevirt/pull/5922/files/8d060997ef64c86ccb03c8480912d429e25b318c..f73036acf559f4cf05f9e1de15e44d1a0b6be667#diff-7a3f9e0461f554b07ac195375eb0f8db7d46579071795a65c459c721b1a8204aR2360).
Note about (3), in case of an error it will update the `migration.Status` that is handled on virt-handler code.

> For example, what happens if the monitor marks the migration as failed, without an abort reason? Is that propagated to the virt-handler and then to the virt-controller which in turn sets this `AbortRequested` to `true`?

In this case it means that the migration was failed and not canceled by the user,
it will propagate to virt-handler as failing migration, vmi.Status.MigrationState.abortRequestd and abortStatus are not in the mix in this case.

Form this line here is how it goes:
Its the virt-controller that set the migration object with AbortRequested = `true`
then virt-handler handles this abort request by sending cancel-migration and finalize-migration requests 
to virt-launcher 
virt-launcher cancelMigration logic runs, which sets the domainSpec with Migration.AbortStatus = 'Aborting'
and cancel Libvirt migration job.
* on the background virt-handler updates the vmi.Status.MigrationState according to the domain migration metadata 
migration-monitor finds that the migration job was canceled and set the domainSpec with
  Completed: true, Failed: true, Phase: Succeeded, Reason: Live migration aborted
* on the background virt-handler updates the vmi.Status.MigrationState ...

[1] https://github.com/kubevirt/kubevirt/pull/5922/files/8d060997ef64c86ccb03c8480912d429e25b318c..f73036acf559f4cf05f9e1de15e44d1a0b6be667#diff-7a3f9e0461f554b07ac195375eb0f8db7d46579071795a65c459c721b1a8204aR2360

ormergi replied:
I edited my last comment with few refinements ^^

ormergi commented on tests/network/vmi_multus.go:
```diff
@@ -1251,6 +1229,20 @@ var _ = SIGDescribe("[Serial]Macvtap", func() {
 	})
 })
 
+func expectInterfaceWithCustomMacToExist(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, macAddress string) {
```
The intent was to expect for an interface by its name to have the given mac, the mac address we configured earlier this is where the custom came from.
On the context of it now, as its own, maybe the name mac no sense..
I think `expectInterfaceToExistByMac` will describe it better

ormergi commented on tests/network/vmi_multus.go:
```diff
@@ -1251,6 +1229,20 @@ var _ = SIGDescribe("[Serial]Macvtap", func() {
+			return err
+		}
+		return checkMacAddress(updatedVMI, interfaceName, macAddress)
+	}, 180*time.Second, 5*time.Second).Should(Succeed())
```
This is a leftover I will parameterize the timeout so it will be 30 seconds on the migration test like before

ormergi replied:
Done

EdDev commented:
> > I see you used `which` embedded in the shell command. I do not understand why it is needed.
> > If the binary is missing, it will give a relevant error, so why do this addition?
> 
> Due to stress tool is invoked as background job if the binary it missing the error will not propagate
> Checking $? after running the command on an image w/o the binary returns '0'

It is confusing that you use `which` when you actually need `test -f <filename>`.
I suggest you either use `test -f` with the expected path of the binary or change the code to express that this `which` addition is making sure just that the binary is there.

ormergi commented:
rebased

ormergi commented on tests/network/vmi_multus.go:
```diff
@@ -1028,21 +1017,10 @@ var _ = Describe("[Serial]SRIOV", func() {
-					return checkMacAddress(updatedVMI, interfaceName, mac)
-				}, 30*time.Second, 5*time.Second).Should(Succeed(),
-					"SR-IOV VF is expected to exist in the guest after migration")
+				By("SR-IOV VF is expected to exist in the guest after migration")
```
Done

ormergi commented:
Changes: 
- Renamed `expectInterfaceWithCustomMacToExist` -> `expectInterfaceToExistByMac`
also parameterized the timeout and expect message (removed `By` usage)

- Added proper check for `stress` tool on `RunStressTest`

- Moved the re-attach "logic" from virt-handler to virt-launcher [[1]](https://github.com/kubevirt/kubevirt/pull/5922/commits/36a9eb7659fc56c1b150081c818afc12544efbef#diff-99bc3793c1cd75cb9339bcc7ff9b949b154bd27c0f48294835642af173314bf5R922).
With the current implementation when a migration is canceled by the client, it will generate an error that eventually
propagate to `asyncMigrate` gorutine [[2]](https://github.com/kubevirt/kubevirt/pull/5922/commits/36a9eb7659fc56c1b150081c818afc12544efbef#diff-99bc3793c1cd75cb9339bcc7ff9b949b154bd27c0f48294835642af173314bf5R893) [[3]](https://github.com/kubevirt/kubevirt/pull/5922/commits/36a9eb7659fc56c1b150081c818afc12544efbef#diff-99bc3793c1cd75cb9339bcc7ff9b949b154bd27c0f48294835642af173314bf5R917-R923) and handled like migration failure.
Thus re-connecting the devices at that point covers both cases where a migration is aborted due to failure or cancel request.

ormergi commented:
/test pull-kubevirt-e2e-kind-1.19-sriov

ormergi commented:
/test pull-kubevirt-e2e-kind-1.19-sriov

ormergi commented:
/retest

ormergi commented:
/test pull-kubevirt-e2e-kind-1.19-sriov

ormergi commented:
/test pull-kubevirt-e2e-kind-1.19-sriov

EdDev commented on tests/migration.go:
```diff
@@ -4,11 +4,14 @@ import (
 	"fmt"
 	"time"
 
+	expect "github.com/google/goexpect"
```
This does not belong to this group (test framework).
I suggest you organize the imports per domain.

ormergi replied:
Done

EdDev commented on tests/network/vmi_multus.go:
```diff
@@ -1251,6 +1231,20 @@ var _ = SIGDescribe("[Serial]Macvtap", func() {
 	})
 })
 
+func expectInterfaceToExistByMac(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, macAddress string, timeout int, expectError string) {
```
The name `expectError` is confusing.
How about `assertMessage` ?

ormergi replied:
Done

EdDev commented on tests/network/vmi_multus.go:
```diff
@@ -1251,6 +1229,20 @@ var _ = SIGDescribe("[Serial]Macvtap", func() {
+func expectInterfaceWithCustomMacToExist(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, macAddress string) {
+	// It may take some time for the VMI interface status to be updated with the information reported by
+	// the guest-agent.
+	Eventually(func() error {
```
What about this?

EdDev commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -918,6 +918,8 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
 			log.Log.Object(vmi).Reason(err).Error("Live migration failed.")
 			migrationErrorChan <- err
+
+			l.hotPlugHostDevices(vmi)
```
I understand that we can not do much if this fails, but it may be useful to log it as an error if it occurs.

ormergi replied:
Done

EdDev reviewed (changes requested):
> sriov migration: Reattach sriov devices when aborted

It is worth explaining shortly why the operation is done this way, especially why it covers all the known scenarios. At least, this is more interesting than the test "how" part.

nit: Please use `SR-IOV` naming format consistently in the commit message.

EdDev commented:
General note for this PR: It currently contains pre-refactoring, a bump and the actual re-attach ability.
For the next time, consider multiple PR/s to allow things to get in quicker, this should help with rebasing overhead.

EdDev commented on tests/migration.go:
```diff
@@ -172,3 +172,23 @@ func RunStressTest(vmi *v1.VirtualMachineInstance, vmsize, stressTimeoutSeconds
 	}
 }
+
+func ExpectMigrationPhaseRunning(virtClient kubecli.KubevirtClient, migration *v1.VirtualMachineInstanceMigration, vmi *v1.VirtualMachineInstance, timeout int) {
```
I find this one odd.
What I see here is a need to "wait" for that migration to start.
It blocks until it detected that it did or failed. If failed, it should come back with an error.

So at the usage, I would expect something like this:
```
Expect(WaitForMigrationPhase(vmim, v1.MigrationRunning).To(Succeed())
```
And there is no need to use `Eventually` then, just a for-loop with a timeout (`wait.PollImmediate`).

ormergi replied:
Looking at this again the function name is misleading,
named `waitForMigrationRunning` (w/o _phase_). 
It checks that the migration is running by checking
both `Migration` object and the `VMI` migration status.
In order to prevent false positive when the VMI status reports the migration is in final state (completed).
Thus it cannot be generic where you can pass the desired phase.

I will rename it and use `wait.PollImmediate` instead of `Eventually`/`Expect`

EdDev commented on tests/network/vmi_multus.go:
```diff
@@ -1022,6 +1022,59 @@ var _ = Describe("[Serial]SRIOV", func() {
 					"SR-IOV VF is expected to exist in the guest after migration")
 			})
+
+			Context("when aborted", func() {
```
You can use the `When` keyword/function instead of `Context`.

ormergi replied:
Done

EdDev commented on tests/network/vmi_multus.go:
```diff
@@ -1022,6 +1022,59 @@ var _ = Describe("[Serial]SRIOV", func() {
+				const vmiStressTimeout = 600
+
+				var migration *v1.VirtualMachineInstanceMigration
+				var migrationUID string
```
Why this is needed? I see that it can just be extrcted from the VMIM.

ormergi replied:
I will remove this and extract it at the tests spec

EdDev commented on tests/network/vmi_multus.go:
```diff
@@ -1022,6 +1022,59 @@ var _ = Describe("[Serial]SRIOV", func() {
+				})
+
+				It("should re-attach SRIOV host-devices to the source VM when failed", func() {
+					updatedVMI, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(vmi.Name, &metav1.GetOptions{})
```
nit: In the other tests, `vmi` was expected to be refreshed to represent the latest status. Why we do this differently now?
Once the "wait" for the phases finishes, it makes sense for it to return with an updated vmi as well IMO.
Anyway, the main reason I even noticed it is because the name here of the vmi variable changed and I was trying to understand why there is a need for two (e.g. do we compare the old and the new?).

Not a must.

ormergi replied:
I added this VMI fetch on this test before adding the annotation because the `Update` command fails because the VMI version is not the latest.

Looking at this again I dont think we need another variable for the update VMI, I will reuse the one that already there.

EdDev commented on tests/network/vmi_multus.go:
```diff
@@ -1022,6 +1022,59 @@ var _ = Describe("[Serial]SRIOV", func() {
+					Expect(err).ToNot(HaveOccurred())
+
+					By("Cause migration failure")
+					// simulate migration target node shutdown by deleting the target pod
+					migrationTargetPod := updatedVMI.Status.MigrationState.TargetPod
+					Expect(virtClient.CoreV1().Pods(migration.Namespace).Delete(context.Background(), migrationTargetPod, metav1.DeleteOptions{})).To(Succeed())
+					tests.WaitForPodToDisappearWithTimeout(migrationTargetPod, targetPodDisposalTimeout)
+
+					tests.ConfirmVMIPostMigrationFailed(virtClient, updatedVMI, migrationUID)
```
Can this be extracted to a function? (e.g. `failMigration()`).

ormergi replied:
Its possible, but since only this test would use it I dont think its necessary for now

EdDev replied:
Functions are not there just for sharing the same code with more than one caller, they also provide readability.
The reader may just want to understand what the test does and not low details on how the failure of the migration is accomplish.

But I won't insist on this.

EdDev commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -865,6 +865,12 @@ func (l *LibvirtDomainManager) migrateHelper(vmi *v1.VirtualMachineInstance, opt
 		}
 	}
 
+	if shouldImmediatelyFailMigration(vmi) {
```
This seems ok to me, but lets hear what the original author of this think about this specific change.
@davidvossel , what do you think? Is this move ok?

davidvossel replied:
yup, it's fine to move this. We just need it to occur before the `dom.MigrateToURI3` function is executed.

ormergi commented:
@EdDev I addressed your code review comments, could you please take another look?

ormergi commented:
/retest

EdDev reviewed (commented):
> tests, SR-IOV migration: Extract SR-IOV VF re-attach asserion to func

typo: asserion

EdDev commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -918,6 +918,10 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
 			migrationErrorChan <- err
+
+			if err := l.hotPlugHostDevices(vmi); err != nil {
+				log.Log.Error("failed to hot-plug host-devices after migration failure")
```
Please log with the object (vmi) and the reason (err).

EdDev commented on tests/migration.go:
```diff
@@ -172,3 +173,29 @@ func RunStressTest(vmi *v1.VirtualMachineInstance, vmsize, stressTimeoutSeconds
+	return wait.PollImmediate(1*time.Second, timeout, func() (bool, error) {
+		migration, err := virtClient.VirtualMachineInstanceMigration(vmim.Namespace).Get(vmim.Name, &metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("failed to get VirtualMachineInstanceMigration %s/%s: %v", migration.Namespace, migration.Name, err)
```
This is true for all the other errors composed here: Are you sure the info you add is not already in the original error?
You should probably add at this level that it failed waiting: "failed waiting for migration to run ...".

EdDev commented on tests/migration.go:
```diff
@@ -172,3 +173,29 @@ func RunStressTest(vmi *v1.VirtualMachineInstance, vmsize, stressTimeoutSeconds
 }
+
+func WaitForMigrationRunning(virtClient kubecli.KubevirtClient, vmim *v1.VirtualMachineInstanceMigration, vmi *v1.VirtualMachineInstance, timeout time.Duration) error {
+	return wait.PollImmediate(1*time.Second, timeout, func() (bool, error) {
```
I think you need to wrap the error (if it occurs) with the context of what failed (see next comment).

EdDev commented on tests/network/vmi_multus.go:
```diff
@@ -1022,6 +1023,57 @@ var _ = Describe("[Serial]SRIOV", func() {
+					tests.ConfirmVMIPostMigrationAborted(virtClient, vmi, string(migration.UID), migrationCompleteTimeout)
+
+					expectInterfaceToExistByMac(virtClient, vmi, mac, reattachSRIOVDeviceTimeout,
+						"SR-IOV VF is expected to exist in the guest after migration was canceled")
```
nit: Okease use the same format in both cancel and failure messages (here you used "in the guest" while in the previous case "on source").

EdDev commented on tests/network/vmi_multus.go:
```diff
@@ -1024,6 +1024,21 @@ var _ = Describe("[Serial]SRIOV", func() {
+
+				migration := tests.NewRandomMigration(vmi.Name, vmi.Namespace)
+				migrationUID := tests.RunMigrationAndExpectFailure(virtClient, migration, tests.MigrationWaitTime)
+				tests.ConfirmVMIPostMigrationFailed(virtClient, vmi, migrationUID)
```
Isn't this covered by `tests.RunMigrationAndExpectFailure` which was prev called?

ormergi replied:
`RunMigrationAndExpectFailure` wait until migration phase is failure 
while `ConfirmVMIPostMigrationFailed` checks that the VMI status.MigrationState reflects migration failure

EdDev replied:
Ohh... one looks at the `VMIM` object while the other on the `VMI`, right?

ormergi commented:
The test that check migration failure during setup using the func tests annotation [[1]](https://github.com/kubevirt/kubevirt/blob/f54e4a5fa74f2cf496e66ad3e7364ff4fc514172/tests/network/vmi_multus.go#L1027)
was stable until this [PR](https://github.com/kubevirt/kubevirt/commit/3276e5449dde860d22bd66319b6151c859fb901c) was merged.
It seems that the VMIM phase wont update to `Failed` and the VMI.Status.Migration state 
wont update with the latest changes from the domain migration metadata.
I verified and the domain spec does have the correct migration metadata after we fail
the migration using the annotation.

There is similar test on SRIOV Live Migration test suite [[2]](https://github.com/kubevirt/kubevirt/blob/main/tests/migration_test.go#L1738), 
I checked it and it passes also with this PR changes on SRIOV provider setup.

@davidvossel @EdDev  maybe you have any idea how that PR could affect it?
In other words why after we re-attach the SRIOV devices after the migration failed
the new domain state wont picked by the `VMController` and `MigrationController`?

Other than that, the other two tests that test migration failure [[3]](https://github.com/kubevirt/kubevirt/blob/31f3d8061b79279787f4dd11dfe4037d54de7392/tests/network/vmi_multus.go#L1050) and cancel while running [[4]](https://github.com/kubevirt/kubevirt/blob/31f3d8061b79279787f4dd11dfe4037d54de7392/tests/network/vmi_multus.go#L1066)
are stable on local env, but wont pass on CI.
As for today the nodes that runs SRIOV lane jobs,
runs other lane jobs as well in parallel - more pressure on the node.
On the previous CI cluster we used to schedule SRIOV jobs **only** on the nodes
that had SRIOV hardware.
I suspect that the pressure on the node affect containerized clusters (KIND infra) 
more than the QEMU VM based clusters.

ormergi commented:
The test that check migration failure during setup using the func tests annotation [[1]](https://github.com/kubevirt/kubevirt/blob/f54e4a5fa74f2cf496e66ad3e7364ff4fc514172/tests/network/vmi_multus.go#L1027)
was stable until https://github.com/kubevirt/kubevirt/pull/5982  was merged [[2]](https://github.com/kubevirt/kubevirt/commit/3276e5449dde860d22bd66319b6151c859fb901c).
It seems that the VMIM phase wont update to `Failed` and the VMI.Status.Migration state 
wont update with the latest changes from the domain migration metadata.
I verified and the domain spec does have the correct migration metadata after we fail
the migration using the annotation.

There is similar test on SRIOV Live Migration test suite [[3]](https://github.com/kubevirt/kubevirt/blob/main/tests/migration_test.go#L1738), 
I checked it and it passes also with this PR changes on SRIOV provider setup.

@davidvossel @vladikr @EdDev maybe you have any idea how that PR could affect it?
In other words why after we re-attach the SRIOV devices after the migration failed
the new domain state wont picked by the VMController and MigrationController?

Other than that, the other two tests that test migration failure [4] and cancel while running [5]
are stable on local env, but wont pass on CI consistently.
As for today the nodes that runs SRIOV lane jobs,
runs other lane jobs as well in parallel.
On the previous CI cluster we used to schedule SRIOV jobs only on the nodes
that had SRIOV hardware.
I suspect that the extra pressure on these nodes affect containerized clusters (KIND infra)
more than the QEMU VM based clusters, because we deploy it in a docker-in-docker pod
it consumes more I/O.
https://github.com/kubernetes-sigs/kind/issues/1922#issuecomment-727620685
AI: 
Check if tests [[4]](https://github.com/kubevirt/kubevirt/blob/31f3d8061b79279787f4dd11dfe4037d54de7392/tests/network/vmi_multus.go#L1050) [[5]](https://github.com/kubevirt/kubevirt/blob/31f3d8061b79279787f4dd11dfe4037d54de7392/tests/network/vmi_multus.go#L1066) are stable when they run in night hours where there is no pressure on CI.
Reduce the amount of jobs these nodes run in parallel and check how it affects   tests stability.
As suggested by @fgimenez

ormergi commented:
/test pull-kubevirt-e2e-kind-1.19-sriov

ormergi commented:
/test pull-kubevirt-e2e-kind-1.19-sriov

ormergi commented:
rebase

ormergi commented:
rebased

davidvossel commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -918,6 +918,10 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
 			log.Log.Object(vmi).Reason(err).Error("Live migration failed.")
 			migrationErrorChan <- err
+
+			if err := l.hotPlugHostDevices(vmi); err != nil {
+				log.Log.Error("failed to hot-plug host-devices after migration failure")
```
since this `hotPlugHostDevices` returns an error, indicating that it could technically fail for some reason, it probably needs to go somewhere that it can be retried.

why not put this logic in SyncVMI(). That way virt-handler will reconcile it and retry if errors occur. I believe we're guaranteed that SyncVMI will get called after a migration failure due to the vmi status update.

ormergi replied:
@davidvossel Hey thanks for the feedback!
`hotPlugHostDevices` is disruptive operation that could have side effects, thus call it only on specific place where we know we want it to be invoked instead of placing it on SyncVMI directly.

It does make sense for me to move it to SyncVMI as you suggest, I prefer to investigate it first and do it on a follow up PR. Moving the logic to SyncVMI changes the behavior of this feature and I want to make sure we do it right :)

davidvossel replied:
We need a solution right out of the gate where this re-hotplug is resilient to failure.

ormergi replied:
@davidvossel When attaching SRIOV VF to the guest fails we can merely raise log/event message about it from kubevirt perspective, I mean we cant fix it on the fly.
Troubleshooting it will probably require a change to the correspond SRIOV PF or on the node
that will probably disrupt the VM.

We didn't want the hotplug on post-migration whether it fails or not to disrupt the migration.
If I got it right, In terms of resilience the VM keep working so I guess it fine, isnt it?

davidvossel replied:
> @davidvossel When attaching SRIOV VF to the guest fails we can merely raise log/event message about it from kubevirt perspective, I mean we cant fix it on the fly.

why not? is there no possibility the failure is transient?

It's fairly common to experience unexpected transient errors while under load, which clear up and successfully execute later on.  If errors aren't handled in an eventually consistent fashion, then those errors would have no chance of being automatically reconciled.

EdDev replied:
@davidvossel , I think your suggestion to reconcile the desired state in regards to the actual state of the domain is ideal, but in this specific case may not be trivial.
- The operation of hot-{un}plug is intrusive to the guest (it mutates the guest), which may result in unknown outcomes to application running there if done too often.
- The action itself is limited to a migration, and we have not generalize it. Therefore, as part of an upgrade, devices are hot-unplug and hot-plugged on the target, but with the current implementation, we do not try to recover.
- The suggestion in the PR is to try and recover the devices in case of migration failure, but still keep it limited to the migration scope only. It is a best-effort operation for now.

If we want to generalize it, and in the reconcile loop to make the hot-{un}plug, then we can move and combine this action of recovery and the action on the target node when migration is finalized.

My suggestion is to bring this forward as a change to the current migration-specific scenario and consider the implications of doing it in that scope.
Keeping the current small scoped domain of migration to one-timer attempts to {un}plug for now.

ormergi commented:
rebased

ormergi commented:
rebased

ormergi commented:
Rebase

EdDev commented:
@ormergi , this PR can be closed, right?

ormergi commented:
> @ormergi , this PR can be closed, right?

Yes this is covered on #6581

ormergi commented:
Close in favor of #6581 
/close
